src/Persistence/AbstractQuery.php

```diff
@@ -281,32 +264,22 @@ public function execute()
         return $this->executeQueryWithDebug(function () {
             $this->withMode();
 
-            // backward compatibility
-            $this->hookOnModel('HOOK_BEFORE_' . $this->getMode(), [$this]);
-
-//          $this->model->hook(self::HOOK_BEFORE_EXECUTE, [$this]);
+            $this->hookOnModel('BEFORE', [$this]);
 
             $result = $this->doExecute();
 
-            // backward compatibility
-            $this->hookOnModel('HOOK_AFTER_' . $this->getMode(), [$this, $result]);
-
-//          $this->model->hook(self::HOOK_AFTER_EXECUTE, [$this, $result]);
+            $this->hookOnModel('AFTER', [$this, $result]);
 
             return $result;
         });
     }
 
     abstract protected function doExecute();
 
-    protected function hookOnModel($name, $args = []): void
+    protected function hookOnModel($when, $args = []): void
     {
-        $hookSpotConst = self::class . '::' . $name;
+        $hookSpotConst = get_class($this->persistence) . '::HOOK_' . $when . '_' . $this->getMode() . '_QUERY';
```

Comment:
no, you must query the constant value! (to not rely on any fixed value)

`string $stage` seems better name vs. `$when`

Reply:
> no, you must query the constant value! (to not rely on any fixed value

This is what I am doing, see 2 rows below. The idea of using `get_class($this->persistence)` is that some persistences can introduce more query modes or hooks

Reply:
> This is what I am doing, see 2 rows below.

I see, already updated my comment... The issue is when you pass bad arg/mode, it will remain undetected...

Reply:
The only way I can think of is declare all constants for all modes (BEFORE_SELECT, AFTER_SELECT and AFTER_DELETE missing now)

Reply:
Yes, I think this is the best approach

Reply:
or use only 2 `BEFORE_EXECUTE_QUERY` and `AFTER_EXECUTE_QUERY`.

Code will be much cleaner then :)

Reply:
you are right, we should use this approach, but dispatch like before (ie. multiple times for ex. in https://user-images.githubusercontent.com/2228672/90612356-37159b00-e208-11ea-9bb9-66d80f5c7074.png) - if we pass mode as the 1st value, then we can check for the first value like before

https://github.com/atk4/core/pull/246 might be harder later, but that is good if solved "properly" :)

I like this approach with `toQuery(string $mode)`, so we can return initialized Query immediatelly. We should have constant for every MODE_ and like other constants, value should start with class name to be sure we use the constant instead of fixed value.